### PR TITLE
Adds a config module for app settings.

### PR DESCRIPTION
**Pull request by hobbes7878**
Alternative to #24.

OK, this creates a new conf module in the app, lightly modeled off django's own settings conf.

The most important part of this implementation to me is it allows all settings to be set _either_ in the django settings module or in the environment, preferring the latter. It also sets all default values in the conf module, rather than in `getattr` statements throughout the app and raises helpful exceptions for any required config a user forgets to set.

Thoughts?

### CODE REVIEW

**Comment by achavez** [comment]
Sorry for the slow response. I'm just now looking at this. Short version: I like the config module idea a lot, I still think some of the usage of env vars is problematic:

- I still think we're reading out of the env in a dangerous way in a couple of spots. Example: You write a test to test S3 publishing. That test writes to a mock S3 and in its teardown wipes the bucket between test runs. If someone has `AWS_ACCESS_KEY_ID` in the env and doesn't properly patch `boto` they'll blow away their entire production bucket during the test tear down (speaking from experience on that one) even if they've properly tried to use `django.test.utils.override_settings`
- There's no way to test behavior that relies on the keys not being set (or a handful of other settings-related conditions) without modifying the environment. For example, it's not possible to write a deterministic test that reaches this branch without modifying the environment: https://github.com/DallasMorningNews/django-chartwerk/blob/3d6743a787eb9531a62e048faddf841be698cdd1/chartwerk/tasks/aws.py#L99

I'd prefer to use Django settings for everything, which I think allows for much more robust and deterministic configuration. I don't love the idea of apps reading out of the env ever, but if we're going to do it I think there are two important changes we should make:

1. Prefer settings over the environment in all cases. Reversing the logic so that the app settings looks in projects settings then the env, I _think_, would fix most of the testing issues above.
2. If we're going to read from the env I think we should namespace everything we're reading from it. So `CHARTWERK_AWS_ACCESS_KEY`, etc. so that if someone is setting that for `boto` or something else they don't have to use that for Chartwerk to allow users to set per-app credentials in their project.

@allanjamesvestal?

**Comment by hobbes7878** [comment]
I'm still not parsing the reluctance to use env vars. I think that's a pretty normal pattern that I think we want to support for a number of reasons. Here's one: at some point I'd like to release this a clonable Heroku app so it comes pre-deployed. At that point, I think it's great if a user can clone and then simply set env vars

**Comment by hobbes7878** [comment]
I'm still not parsing the reluctance to use env vars. I think that's a handy pattern to support that shortens up the config required.

That said, I _don't_ have a problem with preferring settings over env vars. Agree it helps for testing as you suggest.

Name-spacing is happening for everything but AWS creds, so that's an easy fix. I'd still fallback to using default `boto` configuration, though, _unless_ a `CHARTWERK_` namespaced variable is present. And that behavior should be well documented.

Does that work?

**Comment by achavez** [comment]
I think if I had to boil the reluctance down to one thing it's that I think we should always be as explicit at possible, even at the expense of being verbose. So that if we're connecting to AWS on someone's behalf it's through credentials that were explicitly given to us for that purpose through a setting they had to put in code.

I'll absolutely concede that's more a theoretical concern than a technical one, though.

**Comment by allanjamesvestal** [comment]
I agree with @achavez that the current way of reading env vars isn't ideal (and in some cases, like the AWS example above, could even be dangerous to stored data). And I agree that we need to make it so Python code overrides env vars, not the other way around — that's how pretty much everything in python (except boto, which flouts most established best practices anyway and is wrong to do so) works.

I'm always going to be of the opinion that if this is a Django app (and have 'Django' in the branding) it should follow Django's conventions. The people who wrote this framework solved the issue of config with a project-wide `settings.py` file, and doing something inconsistent like app-specific settings just makes things more confusing for future adopters who already know Django.

All we get for handling config differently is that this proposed way makes more conceptual sense to you personally, at the expense of what the Django community at large — and the two of us — would expect. Everything you're talking about, including a one-click deployable Heroku app, is doable in a way that doesn't override that tacit covenant almost all Django apps make with outside developers. (I'll talk about how that one-click Heroku install would work in a second.)

Basically, this is the crux of it: Standards exist for config in Django, and I think [we should adhere to them](https://zen-of-python.info/special-cases-arent-special-enough-to-break-the-rules.html#8).

**Comment by allanjamesvestal** [comment]
Right, so here's the Heroku bit.

We _should_ have a one-click deploy, but it should be in a separate repo.

The reason? That way we can bundle a bare-bones project with `django-chartwerk` as the only non-standard app installed. We can ask for (or generate) all the env vars we'll need in the `app.json` config, and flow those through `./project/settings.py` for a consistent developer experience.

Plus, people who are just plugging this into an app tree don't need a `Procfile` or an `app.json` or any of that other Heroku-specific stuff.

By having a separate deployable repo (which other projects like Sentry do for their Heroku stuff, we can make things pretty effortless to install ("just add S3", plus maybe a couple other required credentials), while keeping that deployment-specific stuff separated from the app logic so as not to confuse users.

Trust me — I'm going through the Sentry deploy path right now, and it's _very_ helpful to know what's Heroku-specific and what's not.

**Comment by achavez** [comment]
Should we sidebar the Heroku piece to a new issue? I really like the idea of a "Deploy to Heroku" button, but I'm thinking it's probably going to outlast the resolution of this PR.

**Comment by allanjamesvestal** [comment]
I'll make a new issue for the Heroku stuff.

**Comment by hobbes7878** [comment]
@allanjamesvestal Have you ever considered a ministry position? That was fairly fire and brimstone, dude.

So, look, I don't think I agree this is all settled orthodoxy. That's why there are packages like [django-appconf](http://django-appconf.readthedocs.io/en/latest/) in the world from people who also write django gud. But I _really_ don't care to argue conventions over function.

So here are the goals:

- Raise helpful errors when required config isn't set
- Set default config values in a single place
- Let me set sensitive values in environment variables (and maybe even encourage me to do so)

I agree with swapping prevalence of python settings over env vars. I prefer to keep the env vars as a direct option, but in the end, if it's going to cause a row, I just don't care as much as y'all.

**Comment by hobbes7878** [comment]
@achavez I'm a week out from whatever we agreed to on this, but see if it looks a'ight to you.
